vault/custom_response_headers.go

```diff
@@ -0,0 +1,193 @@
+package vault
+
+import (
+	"fmt"
+	log "github.com/hashicorp/go-hclog"
+	"net/http"
+	"net/textproto"
+	"strings"
+
+	"github.com/hashicorp/vault/internalshared/configutil"
+)
+
+type ListenersCustomResponseHeadersList struct {
+	CustomHeadersList []*ListenerCustomHeaders
+}
+
+type ListenerCustomHeaders struct {
+	Address string
+	StatusCodeHeaderMap map[string][]*CustomHeader
+	// ConfiguredHeadersStatusCodeMap field is introduced so that we would not need to loop through
+	// StatusCodeHeaderMap to see if a header exists, the key for this map is the headers names
+	ConfiguredHeadersStatusCodeMap map[string][]string
+}
+
+type CustomHeader struct {
+	Name string
+	Value string
+}
+
+func NewListenerCustomHeader(ln []*configutil.Listener, logger log.Logger, uiHeaders http.Header) *ListenersCustomResponseHeadersList {
+
+	if ln == nil {
+		return nil
+	}
+
+	ll := &ListenersCustomResponseHeadersList{}
+
+	for _, l := range ln {
+		lc := &ListenerCustomHeaders{
+			Address: l.Address,
+		}
+		lc.StatusCodeHeaderMap = make(map[string][]*CustomHeader)
+		lc.ConfiguredHeadersStatusCodeMap = make(map[string][]string)
+		for sc, hv := range l.CustomResponseHeaders {
+			var chl []*CustomHeader
+			for h, v := range hv {
+				// Sanitizing custom headers
+				// X-Vault- prefix is reserved for Vault internal processes
+				if strings.HasPrefix(h, "X-Vault-") {
+					logger.Warn("custom headers starting with X-Vault are not valid", "header", h)
+					continue
+				}
+
+				// Checking for UI headers, if any common header exists, we just log an error
+				if uiHeaders != nil {
+					exist := uiHeaders.Get(h)
+					if exist != "" {
+						logger.Warn("found a duplicate header in UI", "header:", h, "Headers defined in the server configuration take precedence.")
+					}
+				}
+
+				// Checking if the header value is not an empty string
+				if v == "" {
+					logger.Warn("header value is an empty string", "header", h, "value", v)
+					continue
+				}
+
+				ch := &CustomHeader{
+					Name: h,
+					Value: v,
+				}
+
+				chl = append(chl, ch)
+
+				// setting up the reverse map of header to status code for easy lookups
+				lc.ConfiguredHeadersStatusCodeMap[h] = append(lc.ConfiguredHeadersStatusCodeMap[h], sc)
+			}
+			lc.StatusCodeHeaderMap[sc] = chl
+		}
+		ll.CustomHeadersList = append(ll.CustomHeadersList, lc)
+	}
+
+	return ll
+}
+
+func (c *ListenersCustomResponseHeadersList) getListenerMap(address string) []*ListenerCustomHeaders {
+	if c.CustomHeadersList == nil {
+		return nil
+	}
+
+	// either looking for a specific listener, or if listener address isn't given,
+	// checking for all available listeners
+	var lch []*ListenerCustomHeaders
+	if address == "" {
+		lch = c.CustomHeadersList
+	} else {
+		for _, l := range c.CustomHeadersList {
+			if l.Address == address {
```

Comment:
If it is presently possible to specify multiple listener config stanzas that reference the same address, that might be a bug.